### DISCUSSION TITLE
Automatic Token KYC with Token Auto Association

### DISCUSSION THREAD

**Discussion by wensheng** [hiero-ledger/hiero-improvement-proposals#239] [Ideas]
Abstract
======
This hip introduces automatic token KYC with token auto association.

Motivation
========
Hedera previously approved and finalized [opt-in token auto association](https://hips.hedera.com/hip/hip-23).  It simplified the flow of a user acquiring a token.  However currently KYC is still required  for a user to receive a token, if the token issuer uses `setKycKey` during token creation.  The current process make certain token activities suck as airdrop cumbersome.

Rationale
=======
When a user has slots available for auto association and an auto association is transacted, KYC should be automatically granted.

User stories
=========

1. will update
2. will update

**Comment by gregscullard** [comment]
@wensheng 

I'm not sure KYC should automatically be granted per your description, this should remain the responsibility of the KYC key holder. For example:
* TokenA is subject to KYC, Alice holds the KYC key.
* Carol has not associated with TokenA and has spare association slots.
* Bob holds tokenA (his KYC was approved by Alice in the past)
* Bob sends TokenA to Carol.
* The transaction should fail since Carol is not KYCd.

We probably should allow this on the other hand

* TokenA is subject to KYC, Alice holds the KYC key.
* Carol has not associated with TokenA and has spare association slots.
* Alice wants to grant KYC to Carol
* The transaction succeeds and results in Carol being automatically associated to TokenA and her KYC being granted.
(The same transaction currently fails since Carol is not associated with the token)

We also should consider extending this to `unfreeze` in the event that a token is frozen by default.
